Update CI

ItsDrike opened the pull request:
This PR introduces various changes to the CI workflows in this repository.

**Note:** Once merged, it will be necessary to update the branch proection rules.

Changes:

- Unify the validation and tox-test worfklows under a single CI workflow, which just imports these as individual jobs.
- Only send the status embed once the unified CI workflow finishes (we don't need to see individual statuses for both workflows)
- Use the more up-to-date syntax for setting values in a step output (using `GITHUB_OUTPUT` env var, instead of `::set-output` in stdout)
- Remove needlessly set values in the `SebastiaanZ/github-status-embed-for-discord` action (the action picks these up by default, no need to set them explicitly)
- Disable fast fail for tox-test matrix jobs, allowing the tests to finish on all versions and platforms. (It can be beneficial to see whether this failure only occured in a single python version/platform, or in all versions/platforms).
- Use the `github` output format for `ruff check`, which provides a lot nicer and easier to read formatting of the found issues, suited specifically for github workflows (uses some specific formatting options that github workflows offer to get nice colors, etc.).
- Create a `tests-done` job as a part of the `tox-test` workflow, which will trigger once all of the matrix test jobs ended, reporting the final status (fails if any of the test jobs failed). This is useful as we can specify it in the branch protection rules, instead of specifying all of the individual tox-test jobs in the matrix, which would require updating these rules every time the required python versions/platforms change.

(This PR is inspired by https://github.com/py-mine/mcproto/pull/300)

PerchunPak commented on .github/workflows/tox-test.yml:
```diff
@@ -35,30 +27,24 @@ jobs:
+  # This job is used purely to provide a workflow status, which we can mark as a
+  # required action in branch protection rules. This is a better option than marking
+  # the tox-test jobs manually, since their names cnange as the supported python
+  # versions change. This job provides an easy single action that can be marked required.
```
This is super clever

ItsDrike commented:
Note: I've updated the branch protection rules. Actually, I moved us to the new branch rulesets instead of the classical protection rules, they provide a bit more options and are just nicer to work with. The ruleset matches our original rules exactly, except for the required workflows, which will now expect the new workflows.

Note 2: Since new workflow names are required, all PRs will need to rebase/merge latest master, to get those workflows to run.
